src/libraries/CalldataDecoder.sol

```diff
@@ -11,35 +11,61 @@ library CalldataDecoder {
 
     error SliceOutOfBounds();
 
-    /// @notice equivalent to SliceOutOfBounds.selector
-    bytes4 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
+    /// @notice mask used for offsets and lengths to ensure no overflow
+    /// @dev no sane abi encoding will pass in an offset or length greater than type(uint32).max
+    ///      (note that this does deviate from standard solidity behavior and offsets/lengths will
+    ///      be interpreted as mod type(uint32).max which will only impact malicious/buggy callers)
+    uint256 constant OFFSET_OR_LENGTH_MASK = 0xffffffff;
+    uint256 constant OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN = 0xffffffe0;
 
-    /// @dev equivalent to: abi.decode(params, (bytes, bytes[])) in calldata
+    /// @notice equivalent to SliceOutOfBounds.selector, stored in least-significant bits
+    uint256 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
+
+    /// @dev equivalent to: abi.decode(params, (bytes, bytes[])) in calldata (requires strict abi encoding)
     function decodeActionsRouterParams(bytes calldata _bytes)
         internal
         pure
         returns (bytes calldata actions, bytes[] calldata params)
     {
         assembly ("memory-safe") {
-            // The offset of the 0th element is 0, which stores the offset of the length pointer of actions array.
-            // The offset of the 1st element is 32, which stores the offset of the length pointer of params array.
-            let actionsPtr := add(_bytes.offset, calldataload(_bytes.offset))
-            let paramsPtr := add(_bytes.offset, calldataload(add(_bytes.offset, 0x20)))
+            // Strict encoding requires that the data begin with:
+            // 0x00: 0x40 (offset to `actions.length`)
+            // 0x20: 0x60 + actions.length (offset to `params.length`)
+            // 0x40: `actions.length`
+            // 0x60: beginning of actions
+
+            // Verify actions offset matches strict encoding
+            let invalidData := xor(calldataload(_bytes.offset), 0x40)
+            actions.offset := add(_bytes.offset, 0x60)
+            actions.length := and(calldataload(add(_bytes.offset, 0x40)), OFFSET_OR_LENGTH_MASK)
 
-            // The length is stored as the first element
-            actions.length := calldataload(actionsPtr)
-            params.length := calldataload(paramsPtr)
+            // Round actions length up to be word-aligned, and add 0x60 (for the first 3 words of encoding)
+            let paramsLengthOffset := add(and(add(actions.length, 0x1f), OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN), 0x60)
```

Comment:
As the `0x60` value is added after applying the mask, the result could actually exceed 32 bites.